pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/PersistentTopicsBase.java

```diff
@@ -4655,6 +4656,74 @@ protected void internalGetLastMessageId(AsyncResponse asyncResponse, boolean aut
                     return null;
         });
     }
+    protected void internalTrimTopic(AsyncResponse asyncResponse, boolean authoritative) {
+        validateTopicOwnershipAsync(topicName, authoritative)
+                .thenCompose(__ -> validateTopicOperationAsync(topicName, TopicOperation.TRIM_TOPIC))
+                .thenCompose(__ -> getTopicReferenceAsync(topicName))
+                .thenCompose(topic -> {
+                    if (topic == null) {
+                        asyncResponse.resume(new RestException(Status.NOT_FOUND,
+                                getTopicNotFoundErrorMessage(topicName.toString())));
+                        return null;
+                    }
+                    if (!(topic instanceof PersistentTopic)) {
+                        log.info("[{}] Trim on a non-persistent topic {} is not allowed", clientAppId(), topicName);
+                        asyncResponse.resume(new RestException(Status.METHOD_NOT_ALLOWED,
+                                "Trim on a non-persistent topic is not allowed"));
+                        return null;
+                    }
+                    if (topicName.isPartitioned()) {
+                        return trimPartitionedTopic(asyncResponse, authoritative);
+                    }
+                    return trimNonPartitionedTopic(asyncResponse, (PersistentTopic) topic);
+                }).exceptionally(ex -> {
+                    // If the exception is not redirect exception we need to log it.
+                    if (!isRedirectException(ex)) {
+                        log.error("[{}] Failed to trim topic {}", clientAppId(), topicName, ex);
+                    }
+                    resumeAsyncResponseExceptionally(asyncResponse, ex);
+                    return null;
+                });
+    }
+
+    @NotNull
+    private static CompletableFuture<Void> trimNonPartitionedTopic(AsyncResponse asyncResponse,
+                                                                   PersistentTopic topic) {
+        CompletableFuture<Void> result = new CompletableFuture<>();
+        Optional.ofNullable(topic.getManagedLedger()).ifPresent(managedLedger ->
+                managedLedger.trimConsumedLedgersInBackground(new CompletableFuture<>()
+                        .thenAccept((v) -> result.complete(null))
+                        .exceptionally(e -> {
+                            result.completeExceptionally(e);
+                            return null;
+                        })));
```

Comment:
I don't think this will works as expected.

```java
managedLedger.trimConsumedLedgersInBackground(new CompletableFuture<>()
                        .thenAccept((v) -> result.complete(null))
                        .exceptionally(e -> {
                            result.completeExceptionally(e);
                            return null;
                        })));
```